onnxruntime/core/optimizer/bias_softmax_fusion.cc

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/common/logging/logging.h"
+#include "core/providers/common.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/bias_softmax_fusion.h"
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/utils.h"
+#include <deque>
+
+#define ILOGF(format_str, ...) \
+  LOGF_DEFAULT(INFO, format_str, ##__VA_ARGS__)
+
+using namespace ONNX_NAMESPACE;
+using namespace ::onnxruntime::common;
+namespace onnxruntime {
+
+// helper to check dimensions match on concrete or symbolic value
+bool operator!=(
+  const ONNX_NAMESPACE::TensorShapeProto_Dimension& lhs,
+  const ONNX_NAMESPACE::TensorShapeProto_Dimension& rhs) {
+    return !(lhs == rhs);
+} 
+
+bool operator==(const ONNX_NAMESPACE::TensorShapeProto_Dimension& lhs, int value) {
+  return utils::HasDimValue(lhs) && lhs.dim_value() == value;
+}
+
+bool operator!=(const ONNX_NAMESPACE::TensorShapeProto_Dimension& lhs, int value) {
+  return !(lhs == value);
+}
+
+void select_input_on_lhs_condition(bool lhs_condition, Node& add_node, NodeArg** input, NodeArg** mask) {
+  if (lhs_condition) {
+    *input = add_node.MutableInputDefs()[0];
+    *mask = add_node.MutableInputDefs()[1];
+  }
+  else {
+    *input = add_node.MutableInputDefs()[1];
+    *mask = add_node.MutableInputDefs()[0];
+  }
+}
+
+Status BiasSoftmaxFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
```

Comment:
nit: this function is well commented, but it's still kind of long. i wonder if splitting it up would help readability.